go/client/tree_test.go

```diff
@@ -112,6 +116,242 @@ func TestBuildTree(t *testing.T) {
+				{path: "b", isSymlink: true, isAbsolute: true, symlinkTarget: "a"},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a/foo.txt", "b"},
+			},
+			want: &client.FileTree{
+				Dirs: map[string]*client.FileTree{
+					"a": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+					"b": &client.FileTree{
+						Files: map[string][]byte{"foo.txt": []byte("foo")},
+					},
+				},
+			},
+		},
+		{
+			desc: "file_exclusions",
+			input: []*inputPath{
+				{path: "a/foo", fileContents: []byte("foo")},
+				{path: "a/b/bar.txt", fileContents: []byte("bar")},
+				{path: "txt/a", fileContents: []byte("a")},
+				{path: "a/b/baz.txt", fileContents: []byte("baz")},
+				{path: "b/bla", fileContents: []byte("bla")},
+				{path: "bla.txt", fileContents: []byte("bla")},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a", "b", "txt", "bla.txt"},
+				InputExclusions: []*command.InputExclusion{
+					&command.InputExclusion{Regex: `txt$`, Type: command.FileInputType},
+				},
+			},
+			want: &client.FileTree{
+				Dirs: map[string]*client.FileTree{
+					"a":   &client.FileTree{Files: map[string][]byte{"foo": []byte("foo")}},
+					"b":   &client.FileTree{Files: map[string][]byte{"bla": []byte("bla")}},
+					"txt": &client.FileTree{Files: map[string][]byte{"a": []byte("a")}},
+				},
+			},
+		},
+		{
+			desc: "dir_exclusions",
+			input: []*inputPath{
+				{path: "b/a", fileContents: []byte("ba")},
+				{path: "a/x", fileContents: []byte("x")},
+				{path: "bla", fileContents: []byte("bla")},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a", "b", "bla"},
+				InputExclusions: []*command.InputExclusion{
+					&command.InputExclusion{Regex: "a$", Type: command.DirectoryInputType},
+				},
+			},
+			want: &client.FileTree{
+				Files: map[string][]byte{"bla": []byte("bla")},
+				Dirs: map[string]*client.FileTree{
+					"b": &client.FileTree{Files: map[string][]byte{"a": []byte("ba")}},
+				},
+			},
+		},
+		{
+			desc: "all_type_exclusions",
+			input: []*inputPath{
+				{path: "b/a", fileContents: []byte("ba")},
+				{path: "a/x", fileContents: []byte("x")},
+				{path: "bla", fileContents: []byte("bla")},
+			},
+			spec: &command.InputSpec{
+				Inputs: []string{"a", "b", "bla"},
+				InputExclusions: []*command.InputExclusion{
+					&command.InputExclusion{Regex: "a$", Type: command.UnspecifiedInputType},
+				},
+			},
+			want: &client.FileTree{},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.desc, func(t *testing.T) {
+			root, err := ioutil.TempDir("", tc.desc)
+			if err != nil {
+				t.Fatalf("failed to make temp dir: %v", err)
+			}
+			defer os.RemoveAll(root)
+			if err := construct(root, tc.input); err != nil {
+				t.Fatalf("failed to construct input dir structure: %v", err)
+			}
+			got, err := client.BuildTreeFromInputs(root, tc.spec)
+			if err != nil {
+				t.Fatalf("BuildTreeFromInputs(%s, %v) = _, %v want _, nil", root, tc.spec, err)
```

Comment:
Should be Errorf since this is not test setup

Reply:
But there's no point testing the next condition if this one is not true, right? I thought Fatalf vs Errorf express this difference, like expectTrue vs assertTrue in Py. http://go/unittestgo#how-do-test-functions-work suggests that, too.

Reply:
I'm basing this off of this guideline: http://go/go-test-comments#keep-going
My understanding is that since you're testing BuildTreeFromInputs, you should check each of its return values and not abort when one is not expected. "When comparing several different properties of a function's output, use t.Error for each of those comparisions".

Reply:
Ah, okay, I interpreted it as the last instruction -- I don't see result and error as orthogonal properties of the output. But I think most people do it the way you suggest. Changed.